### PR TITLE
Update retry in action definition

### PR DESCRIPTION
**Pull request by tsurdilo**
This updates the retry definition of actions. Main reason for update is that currently retry is a single element and it should be an array of retry definitions each matching on the return of the function call.

* Renamed to retry-when and is an array of retry definitions
* Retry definition parameters: match, interval, max-retry, next-state
* Added "default" parameter to Action definition. This defines the state to transition to if the function call fails (error) and none of the retry definitions define a match for it.

### CODE REVIEW

**Comment by tsurdilo** [comment]
ill fix merge conflicts once a decision on this has been made (before merging)

File: workflow/spec/schema/serverless-workflow-schema-01.json [hunk 1]
```diff
@@ -84,25 +84,6 @@
       },
       "required": ["name", "source", "type"]
     },
-    "filter": {
```
Comment (ruromero) [inline workflow/spec/schema/serverless-workflow-schema-01.json]: Why did you remove filters?

File: workflow/spec/schema/serverless-workflow-schema-01.json [hunk 2]
```diff
@@ -154,13 +132,17 @@
           "minimum": 0,
           "description": "Specifies the time period waiting for the events in the EVENTS-EXPRESSION"
         },
-        "retry": {
-          "type": "object",
-          "$ref": "#/definitions/retry",
-          "description": "Specifies how the result from a function is to be handled"
+        "retry-when": {
```
Comment (ruromero) [inline workflow/spec/schema/serverless-workflow-schema-01.json]: A `default` parameter inside an `action` is not easy to be understood as a the default state to go when an error happened.
I suggest you create the following:

* Retry has an array of `match`es and a `default` which will be easily understood as a default match.
* A `match` will then have a `interval`, `max-retry`, `next-state` and instead a `match` I suggest adding a `condition` (see https://github.com/tsurdilo/wg-serverless/pull/1). This will align with the existing way of defining conditions. Currently "match" uses a language expression. We can also use this for event-expressions. WDYT?

```
retry:
  when:
   - match:
     ...
   - match:
     ...
  default: "default-error-state"
```

File: workflow/spec/schema/serverless-workflow-schema-01.json (hunk 1 above)
Comment (tsurdilo) [reply workflow/spec/schema/serverless-workflow-schema-01.json]: this pr was done against master. we had the filters defined in the json schema but not in the spec document. will fix this. sry

File: workflow/spec/schema/serverless-workflow-schema-01.json (hunk 2 above)
Comment (tsurdilo) [reply workflow/spec/schema/serverless-workflow-schema-01.json]: For your example "when" would need to be an array so this can't be described in schema as shown 
Would you be ok with:
```
"retry": {
  "when": [
     {
       ...
     },
     {
       ...
     }
  ]
}
```

Comment (ruromero) [reply workflow/spec/schema/serverless-workflow-schema-01.json]: Yes, that was my point. I oversimplified the example and used yaml. Sorry :smile: 
Do you agree with this?

File: workflow/spec/spec.md [hunk 3]
```diff
@@ -340,8 +340,9 @@ Each event state's event definition includes one or more actions. Let's define t
 | Parameter | Description | Type | Required |
 | --- | --- | --- | --- |
 | [function](#Function-Definition) |Function to be invoked | object | yes |
-| timeout |Max amount of time in seconds to wait for the completion of the function's execution | integer | no |
-| [retry](#Retry-Definition) |Defines if funtion execution needs a retry | object | no |
+| timeout |Max amount of time to wait (in seconds) for the completion of the function's execution | integer | no |
+| [retry-when](#Retry-Definition) |Defines function retries | array | no |
```
Comment (cathyhongzhang) [inline workflow/spec/spec.md]: The name "retry-when" is a little misleading since the semantics is not just about "when". It seems better to just call it"retry" to match the stated description

Comment (tsurdilo) [reply workflow/spec/spec.md]: Ruben suggested a solution in above comment that would make this:
```
"retry": {
   "when": {
     "match": "abc",
...
 },
{
"match": "xyz",
...
}
}
```
the "when" then is assocated with matches so like in language form would be "Retry when match is "abc" and "Retry when match is xyz". 

Would you be ok with that?

File: workflow/spec/schema/serverless-workflow-schema-01.json (hunk 2 above)
Comment (tsurdilo) [reply workflow/spec/schema/serverless-workflow-schema-01.json]: yes

File: workflow/spec/spec.md (hunk 3 above)
Comment (cathyhongzhang) [reply workflow/spec/spec.md]: I see this is closed? Are you going to proceed with this change?

Comment (tsurdilo) [reply workflow/spec/spec.md]: @cathyhongzhang pr is open -- can you let us know if rubens suggestion in comment above is ok with you ?

Comment (cathyhongzhang) [reply workflow/spec/spec.md]: Thanks for checking with me! I think it is better to just use "retry" since 1) the semantics is not just about "when", 2) the syntax is simpler than a two-level syntax. What do you think?

**Comment by tsurdilo** [comment]
rebased and fixed typos

File: workflow/spec/spec.md (hunk 3 above)
Comment (tsurdilo) [reply workflow/spec/spec.md]: sure i have changed to just "retry" without the "when". agree its simpler

**Comment by tsurdilo** [comment]
I have updated to changes requested by both ruben and cathy. This pr has been sitting idle without comments for over a month so I will assume noone has anything else to contribute to it. 

This fixes current issues in master where retry is not array type. 

Will push this pr through as it fixes things. Further changes can be proposed via future prs.
